### DISCUSSION TITLE
Utilize AgentListener inside agent extension

### DISCUSSION THREAD

**Discussion by Cavvar** [open-telemetry/opentelemetry-java-instrumentation#4086] [Q&A] [answered]
Hello everyone!

I was wondering if it is somehow possible to add an AgentListener to an agent extension.

The listener is implemented as follows:
```java
import com.google.auto.service.AutoService;
import io.opentelemetry.instrumentation.api.config.Config;
import io.opentelemetry.javaagent.extension.AgentListener;

@AutoService(AgentListener.class)
public class CustomAgentListener implements AgentListener {

    @Override
    public void beforeAgent(Config config) {
        System.out.println("Hello There;)");
    }
}
```

The java agent is configured like this (Dockerfile.jvm from a Quarkus application):
```docker
ENV JAVA_OPTIONS="-javaagent:/agents/opentelemetry-javaagent-all.jar \
                  -Dotel.traces.exporter=jaeger \
                  -Dotel.exporter.jaeger.endpoint=http://jaeger:14250 \
                  -Dotel.metrics.exporter=none \
                  -Dotel.javaagent.extensions=/agents/custom-agent-extension-0.0.1.jar \
                  -Dquarkus.http.host=0.0.0.0 -Djava.util.logging.manager=org.jboss.logmanager.LogManager"
```

Additionally, I am using a custom `SpanProcessor` and `SpanExporter` which are configured through the `SdkTracerProviderConfigurer` interface.

```java
import com.google.auto.service.AutoService;
import io.opentelemetry.exporter.jaeger.JaegerGrpcSpanExporter;
import io.opentelemetry.javaagent.extension.AgentListener;
import io.opentelemetry.sdk.autoconfigure.spi.SdkTracerProviderConfigurer;
import io.opentelemetry.sdk.trace.SdkTracerProviderBuilder;
import java.util.ServiceLoader;

@AutoService(SdkTracerProviderConfigurer.class)
public class CustomSdkTracerConfigurer implements SdkTracerProviderConfigurer {

    @Override
    public void configure(SdkTracerProviderBuilder sdkTracerProviderBuilder) {
        sdkTracerProviderBuilder.addSpanProcessor(new CustomSpanProcessor(new CustomSpanExporter, true));
    }
}
```

As of right now, the message from the `beforeAgent` hook **is NOT visible**. Am I missing something?

Thanks for any help!

**Comment by mateuszrzeszutek** [comment, marked as answer]
Hey @Cavvar ,
We've recently had this issue https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/4049 that broke javaagent SPIs in extensions. Could you check whether there's a similar exception in your applications logs? (Might need the `-Dotel.javaagent.debug=true` property set for extensive logging)
The issue I linked is already fixed and will be shipped in 1.6.0 by the way.

**Reply by Cavvar** [reply]
Hey @mateuszrzeszutek,

thank you for the quick help!

I wasn't able to find any `AbstractMethodError` but I was able to find other exceptions:
```
java.lang.UnsupportedOperationException: Reflective setAccessible(true) disabled
java.lang.IllegalAccessException: class io.grpc.netty.shaded.io.netty.util.internal.PlatformDependent0$6 cannot access class jdk.internal.misc.Unsafe (in module java.base) because module java.base does not export jdk.internal.misc to unnamed module
Suppressed: java.lang.UnsatisfiedLinkError: no io_grpc_netty_shaded_netty_transport_native_epoll_x86_64 in java.library.path
```

Are these related to #4049?

**Reply by Cavvar** [reply]
Here the full log output:
[bug-agent-listener.txt](https://github.com/open-telemetry/opentelemetry-java-instrumentation/files/7144627/bug-agent-listener.txt)

**Reply by trask** [reply]
hey @Cavvar, I think you will run into #4049, but since you aren't getting that exception I wonder if you are running into something else first. you can try with the [latest snapshot](https://oss.sonatype.org/content/repositories/snapshots/io/opentelemetry/javaagent/opentelemetry-javaagent/1.6.0-SNAPSHOT/opentelemetry-javaagent-1.6.0-20210910.195802-72-all.jar) to eliminate #4049 as the issue. if your AgentListener is still not being called, can you publish a small repro of your extension that we can look at?

**Reply by Cavvar** [reply]
Hello @trask. I used the snapshot and was able to see the logging message from inside the `beforeAgent` hook. Thanks for the help!

**Reply by Yahire3388** [reply]
Hi @Cavvar , @trask 

can you share the snapshot again? it is not available at the link.

Thanks

**Reply by mateuszrzeszutek** [reply]
Try the [1.13.0-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/io/opentelemetry/javaagent/opentelemetry-javaagent/1.13.0-SNAPSHOT/opentelemetry-javaagent-1.13.0-20220317.033940-21.jar)

**Reply by Yahire3388** [reply]
Hi @mateuszrzeszutek @trask 
Tried, but it doesn't seem to add any custom traces. I am trying to export the traces directly to database by adding Mongo db connection code in agent extension from Span processor or by span exporter class. 

Thanks
